rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+
+```cpp
+class Split {
+ public:
+  virtual std::string DebugString() const = 0;
+  // Methods to support being used as a Variant tensor.
+  virtual std::string TypeName() const = 0;
+  virtual void Encode(VariantTensorData* data) const = 0;
+  virtual bool Decode(const VariantTensorData& data) = 0;
+};
+```
+
+To iterate over splits for a dataset, we will use a new
+`DatasetBase::MakeSplitGenerator()` method. This method creates a
+`SplitGenerator`, which is responsible for generating all of the splits for the
+dataset. We use an intermediate `SplitGenerator` object instead of generating
+splits directly because there could be a large number of splits, and the
+`SplitGenerator` gives us as way to tune split size in response to pipeline
+performance.
+
+```cpp
+class SplitGenerator {
+ public:
+  virtual Status GetNext(std::unique_ptr<Split>* split,
+                         bool* end_of_splits) = 0;
+  // Instructs the SplitGenerator to adjust the size of future splits by the
+  // specified percent. 100% means no change, 50% means half-sized splits, and
+  // 200% means double-sized splits. The SplitGenerator will make a best effort
+  // to incorporate the feedback when creating splits.
+  virtual void AdjustSplitSize(int percent) = 0;
+};
+```
+
+It is tempting to process each split independently, but this would cause issues
+when splits are small. tf.data pipelines need to populate internal buffers for
+shuffling, prefetching, and batching. If we use a separate pipeline to process
+each split, our shuffling will be lower quality, we will have performance jitter
+as we keep needing to refill prefetch buffers from scratching, and we will
+produce many more partial batches (each split might not even have enough data to
+fill a full batch). To avoid these issues, we use a small number of tasks, where
+each task processes many splits as a single pipeline.
+
+To enable processing of multiple splits in a dataset, we will add an optional
+`SplitProvider` field to the `IteratorContext` passed to
+`IteratorBase::Initialize`. The `SplitProvider` produces splits which tell the
+iterator what source data to iterate over. For example, if splits are
+represented by filenames, and a SplitProvider produces `["file1", "file6",
+"file11"]`, an iterator initialized by that `SplitProvider` should process those
+three files only.
+
+```cpp
+class SplitProvider {
+ public:
+  virtual Status GetNext(std::unique_ptr<Split>* split,
+                         bool* end_of_splits) = 0;
+};
+```
+
+When processing datasets, tf.data service workers will use `SplitProvider`s
+which provide splits by querying the tf.data service master for which splits to
+process. A few splits will be prefetched to hide the latency of needing to
+request a new split from the master.
+
+#### Supported Datasets
+
+Not all dataset sources and transformations are easily splittable. For example,
+`take`, `skip`, and `scan` require a global view of the dataset to produce
```

Comment:
How do take and skip require global views?

Reply:
take and skip use the global order of elements to determine which elements to produce. With splitting, the order is blurred so that it isn't clear how to compare the order of elements produced from different splits.